EDU-2838: update the service health and worker health docs (#3024)
temporalio/documentation: docs/production-deployment/cloud/worker-health.mdx

  - temporal worker resources
---

> This page is a guide to monitoring a Temporal Worker fleet, which covers these scenarios:

Reviewer: ```suggestion
This page is a guide to monitoring a Temporal Worker fleet and covers the following scenarios:
```

I really like the expansion of the sentence.

- **Temporal Cloud metric**: [temporal_cloud_v0_poll_success_count](/production-deployment/cloud/metrics/reference#temporal_cloud_v0_poll_success_count)
- **Temporal Cloud metric**: [temporal_cloud_v0_poll_success_sync_count](/production-deployment/cloud/metrics/reference#temporal_cloud_v0_poll_success_sync_count)

### Schedule To Start latency

The Schedule-To-Start metric represents how long Tasks are staying, unprocessed, in the Task Queues.
But differently, it is the time between when a Task is enqueued and when it is picked up by a Worker.
This time being long (likely) means that your Workers can't keep up — either increase the number of Workers (if the host load is already high) or increase the number of pollers per Worker.

> If your Schedule-To-Start latency alert fires or appears high, cross-check this with the [Sync Match Rate](#sync-match-rate) to determine if you need to change your Worker or fleet, or if you need to contact Temporal Cloud support.

Reviewer: ```suggestion
If your Schedule-To-Start latency alert triggers or is high, check the [Sync Match Rate](#sync-match-rate) to decide if you need to adjust your Worker or fleet, or contact Temporal Cloud support.
If your Sync Match Rate is low, contact [Temporal Cloud support](/cloud/support#support-ticket).
```

**Target**

> This latency should be a very low value -- close to zero.

Reviewer: ```suggestion
This latency should be very low, close to zero.
Any higher value indicates a bottleneck.
```

Anything else indicates bottlenecking.

### Sync Match Rate {#sync-match-rate}

The Sync Match Rate measures the rate of Tasks that can be delivered to Workers without having to be persisted (Workers are up and available to pick them up) to the rate of all delivered Tasks.

**Calculate Sync Match Rate**

```
temporal_cloud_v0_poll_success_sync_count ÷ temporal_cloud_v0_poll_success_count = N

> ```

Reviewer: Isn't $$ the indicator for in-line equations? This may be intentional math.

Author: Yes, but it broke the page width and wasn't really rendering properly, nor was it used elsewhere on this same page, so I think it was a non-working convention.

> #### Prometheus query samples

Reviewer: Just confirming that since this isn't a Temporal Query, that it's lower case. Putting it into words so you know I looked, considered, and agreed.

  - Golang: [Workflows](/develop/go/core-application#maxconcurrentworkflowtaskpollers)
- Doing some combination of the above

**Temporal Cloud bottleneck**

If the Schedule-To-Start latency is _high_ and the Sync Match Rate is also _low_, Temporal Cloud could very well be the bottleneck and you should reach out via support channels for us to confirm.

> :::warning

Reviewer: For some reason in my head, I was sure this was "caution" but I looked it up in the style guide and it's warning.

> If you are setting the [Schedule-To-Start Timeout](/encyclopedia/detecting-activity-failures#schedule-to-start-timeout) value in your Activity Options, it will skew your observations.

Reviewer: ```suggestion
Setting the [Schedule-To-Start Timeout](/encyclopedia/detecting-activity-failures#schedule-to-start-timeout) in your Activity Options can skew your observations.
Avoid setting a Schedule-To-Start Timeout when profiling latency.
```

> **How to detect misconfigured Workers.**

Reviewer: This is that example I was referring to above. It's really inconsistent where the bolded questions happen and where they don't right now. We should adopt and enforce one or the other.

> The configuration of your Workers can negatively affect the efficiency of your Task processing as well.

Reviewer: ```suggestion
Worker configuration can affect Task processing efficiency.
```

Increase the `maxConcurrentWorkflowTaskExecutionSize` and `maxConcurrentActivityExecutionSize` values and keep an eye on your Worker resource metrics (CPU utilization, etc) to make sure you haven't created a new issue.

### Configure Sticky Execution Cache {#configure-sticky-cache}

> A Sticky Execution is a when a Worker Entity caches the Workflow Execution Event History and creates a dedicated Task Queue to listen on.

Reviewer: ```suggestion
Sticky Execution means that a Worker caches a Workflow Execution Event History and creates a dedicated Task Queue to listen on.
It significantly improves performance because the Temporal Service only sends new events to the Worker instead of entire Event Histories.
```
